src/python/pants/engine/internals/target_adaptor.py

```diff
@@ -104,10 +104,19 @@ def __init__(
     ) -> None:
         self.type_alias = type_alias
         self.name = name
-        self.kwargs = kwargs
+        self.kwargs = FrozenDict.deep_freeze(kwargs)
```

Comment:
This requires all field values to be hash-able from the BUILD file. Are we sure this is true?

Before, we rely on the `Field` class to transform the BUILD file value into the immutable hashable type in `Field.compute_value`.

I worry this could be a breaking change.

Reply:
It had better be true, since we were implicitly assuming it was (via the assumption that the `__hash__` on `SyntheticAddressMaps` was correct)...

In other words, if that `__hash__` was correct then it would have choked on an unhashable `AddressMap`, which would have choked on an unhashable `TargetAdaptor`, which would have choked on an unhashable kwargs... 

So the issue is that we're now enforcing something that we were always assuming, but failed to enforce. So yeah, it might break things, but in a necessary way, I guess?

Also, note that `deep_freeze()` will always work, as it just turns dicts into FrozenDicts and lists/sets into tuples. It has nothing to say about the values inside those structures.  So if something blows up, it'll be when we actually try to compute the hash...

I will add this to the release notes, but I imagine that in practice this should be OK, since the values inside those structures are overwhelmingly likely to be ints/floats/bools/strs no?

Reply:
> Also, note that deep_freeze() will always work, as it just turns dicts into FrozenDicts and lists/sets into tuples. It has nothing to say about the values inside those structures. So if something blows up, it'll be when we actually try to compute the hash...

[`Parametrize`](https://github.com/pantsbuild/pants/blob/3f1793623c56e7bf1e498495530757603a55250b/src/python/pants/engine/internals/parametrize.py#L40) might be an issue since it is a frozen dataclass but has several members which mutate it, and instances of the class can show up in the `kwargs`.

Reply:
`parametrize` freezes its kwargs, but assumes without proof that its positional args are strs. That is what we document, both on the docsite and in the type annotations, but nothing actually checks, apparently...

Reply:
> Also, note that deep_freeze() will always work, as it just turns dicts into FrozenDicts and lists/sets into tuples. It has nothing to say about the values inside those structures. So if something blows up, it'll be when we actually try to compute the hash...

Hrrm.. yea. And we compute the hash in `FrozenDict.__init__`, so `deep_freeze()` will fail, if there is any unhashable data in there. The only unhashable values that are treated are `dict`, `list` and `set`, which I agree will be the absolute majority of values. But any other unhashable values than those, will blow up unless they implement a custom `__hash__` method.

I am not aware of any object types pants provides to the BUILD file scope that are unhashable (the two main ones that comes to mind are `python_artifact` and `http_source`, and both of these are hashable.) but any plugin may provide a data type that isn't hashable. And blowing up in that case is the right thing to do, so with the added note in the changelog, I'm happy.

> I will add this to the release notes, but I imagine that in practice this should be OK, since the values inside those structures are overwhelmingly likely to be ints/floats/bools/strs no?

👍🏽 
yes, it should be OK for 99.9% (arbitrary value, obviously) of the cases, but there's no guarantee it won't fail. So calling this out is the right thing to do, imo.

But given that, I think we could improve the error message if this happens, as we know the origin of the bad data, it would just be cruel not to convey it. I think this is worth while to add, given that this is not just something that may hit current users/setups but also future ones, and it is a mistake more likely to be made by new users/plugin authors that may not be accustomed to working with immutability, which could yield a potentially cryptic unhelpful error message.

Reply:
Added the description of origin to the already pretty informative FrozenDict TypeError.